common/workunit/workunit.cpp

```diff
@@ -13383,17 +13383,52 @@ extern WORKUNIT_API void addTimeStamp(IWorkUnit * wu, StatisticScopeType scopeTy
     wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), scopeType, scopestr, kind, NULL, getTimeStampNowValue(), 1, 0, StatsMergeAppend);
 }
 
+static double getCpuSize(const char *resourceName)
+{
+    const char * cpuRequestedStr = queryComponentConfig().queryProp(resourceName);
+    if (!cpuRequestedStr)
+        return 0.0;
+    char * endptr;
+    double cpuRequested = strtod(cpuRequestedStr, &endptr);
+    if (cpuRequested < 0.0)
+        return 0.0;
+    if (*endptr == 'm')
+        cpuRequested /= 1000;
+    return cpuRequested;
+}
+
+static double getCostCpuHour()
+{
+    const char * costCpuHourStr = queryGlobalConfig().queryProp("cost/@perCpu");
+    if (!costCpuHourStr)
+        return 0.0;
+    double costCpuHour = strtod(costCpuHourStr, nullptr);
+    if (costCpuHour < 0.0)
+        return 0.0;
+    return costCpuHour;
+}
+
+extern WORKUNIT_API cost_type getMachineCostRate()
+{
+    return getCostCpuHour() * getCpuSize("resources/@cpu") ;
+};
+
+extern WORKUNIT_API double getThorManagerRate()
+{
+    return getCostCpuHour() * getCpuSize("managerResources/@cpu");
+}
+
+extern WORKUNIT_API double getThorWorkerRate()
+{
+    return getCostCpuHour() * getCpuSize("workerResources/@cpu");
+}
+
 extern WORKUNIT_API cost_type calculateThorCost(unsigned __int64 ms, unsigned clusterWidth)
 {
-    IPropertyTree *costs = queryCostsConfiguration();
-    if (costs)
-    {
-        cost_type thor_master_rate = money2cost_type(costs->getPropReal("thor/@master"));
-        cost_type thor_slave_rate = money2cost_type(costs->getPropReal("thor/@slave"));
+    const cost_type managerRate = money2cost_type(getThorManagerRate());
+    const cost_type workerRate = money2cost_type(getThorWorkerRate());
 
-        return calcCost(thor_master_rate, ms) + calcCost(thor_slave_rate, ms) * clusterWidth;
-    }
-    return 0;
+    return calcCost(managerRate, ms) + calcCost(workerRate, ms) * clusterWidth;
```

Comment:
I'm slightly concerned about rounding errors.  With the cost you have defined, you would get no cost for items < ~30ms.  Probably not a big loss, but I wonder if we should change the scaling factor for money to cost from 1E6 to 1E10, or even higher.

Reply:
It would be best to use floating point until the last minute and then convert to a cost.

Reply:
I'm keeping it as a double and converting at the last minute.  Do you think it is still necessary to have a higher scaling factor?